### ISSUE TITLE
[Bug Report]Unexpected result torch.abs(x - y) >= 0 while x == y

### ISSUE THREAD

**Issue by twotwoiscute** [pytorch/pytorch#79262]
### 🐛 Describe the bug

The following code really confuses me , the cond = n < beta is never true, However when I do backward pass input grad is NaN, I believe this is because 0.5 * n**2 / beta, I would like to know how is this possible ? Thanks

```
import torch 
torch.__version__
'1.8.0+cpu'
input = torch.tensor([[-0.6205, 0.225],[-0.8916, 0.4492]], requires_grad = True) 
target = torch.tensor([[-0.6205, 0.225],[-0.8916, 0.4492]]) 
beta = 0 
n = torch.abs(input - target) 
#Since input == target so the cond is never true.
cond = n < beta 
loss = torch.where(cond, 0.5 * n**2 / beta, n - 0.5 * beta)
loss = loss.sum() 
loss.backward() 
print(input.grad)
tensor([[nan, nan],
       [nan, nan]])


### Versions

torch version : '1.8.0+cpu'

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7

**Comment by vadimkantorov** [comment]
Probably torch.where would still produce nan gradients even if the condition is never true. Related discussion: https://github.com/pytorch/pytorch/issues/31829
